### ISSUE TITLE
Error (criu/mem.c:726): Unable to remap a private vma: Invalid argument

### ISSUE THREAD

**Issue by avagin** [xemul/criu#322]
```
(00.410614)      4: cg: Cgroups 1 inherited from parent
(00.410858)      4: Opened local page read 3 (parent 0)
(00.410961)      4:     premap 0x00000000400000-0x00000000406000 -> 00007fe65badf000
(00.410981)      4:     premap 0x00000000605000-0x00000000606000 -> 00007fe65bae5000
(00.410997)      4:     premap 0x00000000606000-0x00000000607000 -> 00007fe65bae6000
(00.411013)      4:     premap 0x000000025a0000-0x000000025c1000 -> 00007fe65bae7000
(00.411036)      4: Error (criu/mem.c:726): Unable to remap a private vma: Invalid argument
(00.412779)      1: Error (criu/cr-restore.c:1465): 4 exited, status=1
(00.412819)      1: Restoring resources
(00.412850)      1: Opening fdinfo-s
(00.412886)      1:             Create fd for 0
(00.412905)      1:             Create fd for 1
(00.412913)      1:                     Going to dup 1 into 2
(00.412923)      1:     Receive fd for 2
(00.413215)      1: `- render 1 iovs (0x7f1b8440f000:12288...)
(00.413251)      1: Restore via sigreturn
```

```
14245 pts/0    T      0:00  |           |       \_ ./zdtm_ct zdtm.py
14247 pts/0    S      0:00  |           |           \_ python2 zdtm.py
14249 pts/0    T      0:00  |           |               \_ python2 zdtm.py
14294 pts/0    S      0:00  |           |                   \_ ../criu/criu restore -o restore.log -D dump/zdtm/static/shm/31/1 -v4 --pidfile /root/git/criu/test/zdtm/static/shm.pid --root /
14296 pts/0    S      0:00  |           |                       \_ ../criu/criu restore -o restore.log -D dump/zdtm/static/shm/31/1 -v4 --pidfile /root/git/criu/test/zdtm/static/shm.pid --ro
14310 pts/0    S      0:00  |           |                       |   \_ ../criu/criu restore -o restore.log -D dump/zdtm/static/shm/31/1 -v4 --pidfile /root/git/criu/test/zdtm/static/shm.pid 
14297 ?        Ss     0:00  |           |                       \_ ../criu/criu restore -o restore.log -D dump/zdtm/static/shm/31/1 -v4 --pidfile /root/git/criu/test/zdtm/static/shm.pid --ro
14481 pts/0    R+     0:00  |           \_ ps axf
```

kernel: 4.12-rc6

**Comment by avagin** [comment]
```
146   mremap(0x7f0aab85c000, 1822720, 1822720, MREMAP_MAYMOVE|MREMAP_FIXED, 0x7f0aab70c000) = -1 EINVAL (Invalid argument) <0.000018>
```

**Comment by avagin** [comment]
commit 1be7107fbe18eed3e319a6c3e83c78254b693acb
Author: Hugh Dickins <hughd@google.com>
Date:   Mon Jun 19 04:03:24 2017 -0700

    mm: larger stack guard gap, between vmas

**Comment by cyrillos** [comment]
One of the key moments in the patch:
---
diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
index f0c8b33..520802d 100644
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -300,11 +300,7 @@ show_map_vma(struct seq_file *m, struct vm_area_struct *vma, int is_pid)
 
        /* We don't show the stack guard page in /proc/maps */
        start = vma->vm_start;
-       if (stack_guard_page_start(vma, start))
-               start += PAGE_SIZE;
        end = vma->vm_end;
-       if (stack_guard_page_end(vma, end))
-               end -= PAGE_SIZE;
 
        seq_setwidth(m, 25 + sizeof(void *) * 6 - 1);
        seq_printf(m, "%08lx-%08lx %c%c%c%c %08llx %02x:%02x %lu ",

**Comment by Fabian-Gruenbichler** [comment]
just as a heads up: this is the fix for CVE-2017-1000364 aka "The Stack Clash" - this is not only in -rc6, but released in most distros as security fix today (saw this flying by on LKML, but didn't want to spam the list).

**Comment by cyrillos** [comment]
Yeah, we know. Thanks!

**Comment by cyrillos** [comment]
https://lists.openvz.org/pipermail/criu/2017-June/thread.html

**Comment by tswift242** [comment]
Do we know which criu versions are affected by this? I'm using criu 2.12.1 and want to know if pulling in this kernel fix will break checkpoint/restore.

**Comment by cyrillos** [comment]
the patches are flying by now, and the kernel is not yet stabilized itself. that said I think next 3.x version will carry the fix.

**Comment by xemul** [comment]
@tswift242 , any criu is affected :) This is "we don't break userspace" rule violation by the kernel :(
@cyrillos , do kernel guys insist on fixing this on criu side only?

**Comment by cyrillos** [comment]
@xemul Well, the stack-guard engine got reworked to fix vulnerability so procfs output get changed and requires criu to be patched accordingly. There is no option for us.

**Comment by tswift242** [comment]
@xemul @cyrillos I just tried upgrading to kernel 4.11.7 and checkpoint/restore continued working for me (using criu 2.12.1). That release contains the upstream kernel commit 1be7107fbe18eed3e319a6c3e83c78254b693acb. Would you expect this to fail? Did the kernel do something to avoid breaking criu? Or are there certain situations in which the issue talked about in this thread doesn't reproduce (e.g. certain kernel versions, distros, applications, etc.)? I'm trying to figure out if it's safe to do this upgrade while using C/R.

**Comment by cyrillos** [comment]
The vanilla kernel is safe, while we're seeing wrong patch being pushed into some distros like ubuntu.

**Comment by xemul** [comment]
Fixed in 3.2.1
